File: sessao-5/tutorial-10-refatoracao-assistida/exemplos/importacao_revisado.py
"""
VERSÃO REVISADA — importação de clientes a partir de CSV após refatoração assistida
Referência: Clean Code, Cap. 3 (Funções)

Problemas corrigidos em relação à versão gerada:
  - Função monolítica dividida em responsabilidades únicas
  - ler_linhas: lê e descarta o cabeçalho
  - validar_cliente: verifica campos obrigatórios e formato de e-mail
  - converter_cliente: monta o dicionário com campos normalizados
  - importar_clientes: orquestra as três funções acima
  - Nomes descritivos em português para todos os identificadores

Nota didática: o comportamento é idêntico ao da versão gerada — apenas
a estrutura interna foi reorganizada em passos pequenos e verificáveis.

Execute: python3 sessao-5/tutorial-10-refatoracao-assistida/exemplos/importacao_revisado.py
"""

from __future__ import annotations

SEPARADOR_CSV = ","
INDICE_NOME = 0
INDICE_EMAIL = 1
INDICE_CIDADE = 2
TOTAL_CAMPOS_ESPERADO = 3


def ler_linhas(conteudo_csv: str) -> list[str]:
    """Retorna as linhas de dados do CSV, descartando o cabeçalho."""
    linhas = conteudo_csv.strip().split("\n")
    return linhas[1:]


def validar_cliente(campos: list[str]) -> bool:
    """Retorna True se os campos contiverem nome, e-mail válido e cidade."""
    if len(campos) != TOTAL_CAMPOS_ESPERADO:
        return False
    nome = campos[INDICE_NOME].strip()
    email = campos[INDICE_EMAIL].strip()
    return bool(nome) and bool(email) and "@" in email


def converter_cliente(campos: list[str]) -> dict:
    """Converte campos CSV em dicionário de cliente normalizado."""
    return {
        "nome": campos[INDICE_NOME].strip(),
        "email": campos[INDICE_EMAIL].strip(),
        "cidade": campos[INDICE_CIDADE].strip(),
    }


def importar_clientes(conteudo_csv: str) -> list[dict]:
    """Lê, valida e converte cada linha do CSV em dicionário de cliente."""
    clientes = []
    for linha in ler_linhas(conteudo_csv):
        campos = linha.split(SEPARADOR_CSV)
        if validar_cliente(campos):
            clientes.append(converter_cliente(campos))
    return clientes


# ─── Execução de demonstração ─────────────────────────────────────────────────

if __name__ == "__main__":
    dados = (
        "nome,email,cidade\n"
        "Ana Silva,ana@exemplo.com,São Paulo\n"
        "Carlos,carlos-invalido,Rio de Janeiro\n"
        "Beatriz Santos,beatriz@exemplo.com,Curitiba\n"
        ",email@vazio.com,Belo Horizonte\n"
        "Diego Lima,diego@exemplo.com,Porto Alegre\n"
    )

    clientes = importar_clientes(dados)
    print(f"Clientes importados: {len(clientes)}")
    for cliente in clientes:
        print(f"  {cliente['nome']} <{cliente['email']}> — {cliente['cidade']}")

File: sessao-5/tutorial-10-refatoracao-assistida/exemplos/importacao_revisado.ts
/**
 * VERSÃO REVISADA — importação de clientes a partir de CSV após refatoração assistida
 * Referência: Clean Code, Cap. 3 (Funções)
 *
 * Problemas corrigidos em relação à versão gerada:
 *   - Função monolítica dividida em responsabilidades únicas
 *   - lerLinhas: lê e descarta o cabeçalho
 *   - validarCliente: verifica campos obrigatórios e formato de e-mail
 *   - converterCliente: monta o objeto com campos normalizados
 *   - importarClientes: orquestra as três funções acima
 *   - Nomes descritivos em português para todos os identificadores
 *
 * Nota didática: o comportamento é idêntico ao da versão gerada — apenas
 * a estrutura interna foi reorganizada em passos pequenos e verificáveis.
 *
 * Execute: npx ts-node sessao-5/tutorial-10-refatoracao-assistida/exemplos/importacao_revisado.ts
 */

const SEPARADOR_CSV = ",";
const INDICE_NOME = 0;
const INDICE_EMAIL = 1;
const INDICE_CIDADE = 2;
const TOTAL_CAMPOS_ESPERADO = 3;

interface Cliente {
  nome: string;
  email: string;
  cidade: string;
}


function lerLinhas(conteudoCsv: string): string[] {
  const linhas = conteudoCsv.trim().split("\n");
  return linhas.slice(1);  // descarta cabeçalho
}


function validarCliente(campos: string[]): boolean {
  if (campos.length !== TOTAL_CAMPOS_ESPERADO) {
    return false;
  }
  const nome = campos[INDICE_NOME].trim();
  const email = campos[INDICE_EMAIL].trim();
  return nome.length > 0 && email.length > 0 && email.includes("@");
}


function converterCliente(campos: string[]): Cliente {
  return {
    nome: campos[INDICE_NOME].trim(),
    email: campos[INDICE_EMAIL].trim(),
    cidade: campos[INDICE_CIDADE].trim(),
  };
}


function importarClientes(conteudoCsv: string): Cliente[] {
  const clientes: Cliente[] = [];
  for (const linha of lerLinhas(conteudoCsv)) {
    const campos = linha.split(SEPARADOR_CSV);
    if (validarCliente(campos)) {
      clientes.push(converterCliente(campos));
    }
  }
  return clientes;
}


// ─── Execução de demonstração ─────────────────────────────────────────────────

const dados =
  "nome,email,cidade\n" +
  "Ana Silva,ana@exemplo.com,São Paulo\n" +
  "Carlos,carlos-invalido,Rio de Janeiro\n" +
  "Beatriz Santos,beatriz@exemplo.com,Curitiba\n" +
  ",email@vazio.com,Belo Horizonte\n" +
  "Diego Lima,diego@exemplo.com,Porto Alegre\n";

const clientes = importarClientes(dados);
console.log(`Clientes importados: ${clientes.length}`);
for (const cliente of clientes) {
  console.log(`  ${cliente.nome} <${cliente.email}> — ${cliente.cidade}`);
}
